mssql/index.d.ts

```diff
@@ -1,91 +1,96 @@
 // Type definitions for mssql 3.3
 // Project: https://www.npmjs.com/package/mssql
-// Definitions by: COLSA Corporation <http://www.colsa.com/>, Ben Farr <https://github.com/jaminfarr>, Vitor Buzinaro <https://github.com/buzinas>, Matt Richardson <https://github.com/mrrichar/>
+// Definitions by: COLSA Corporation <http://www.colsa.com/>, Ben Farr <https://github.com/jaminfarr>, Vitor Buzinaro <https://github.com/buzinas>, Matt Richardson <https://github.com/mrrichar/>, Jørgen Elgaard Larsen <https://github.com/elhaard/>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 
 /// <reference types="node" />
 
 
 import events = require('events');
+export interface ISqlType {
+  type: ISqlTypeFactory;
+}
+export interface ISqlTypeWithNoParams extends ISqlType { type: ISqlTypeFactoryWithNoParams }
+export interface ISqlTypeWithLength extends ISqlType { type: ISqlTypeFactoryWithLength; length: number }
+export interface ISqlTypeWithScale extends ISqlType { type: ISqlTypeFactoryWithScale; scale: number }
+export interface ISqlTypeWithPrecisionScale extends ISqlType { type: ISqlTypeFactoryWithPrecisionScale; precision: number, scale: number }
+export interface ISqlTypeWithTvpType extends ISqlType { type: ISqlTypeFactoryWithTvpType; tvpType: any }
 
-type sqlTypeWithNoParams = { type: sqlTypeFactoryWithNoParams }
-type sqlTypeWithLength = { type: sqlTypeFactoryWithLength, length: number }
-type sqlTypeWithScale = { type: sqlTypeFactoryWithScale, scale: number }
-type sqlTypeWithPrecisionScale = { type: sqlTypeFactoryWithPrecisionScale, precision: number, scale: number }
-type sqlTypeWithTvpType = { type: sqlTypeFactoryWithTvpType, tvpType: any }
+export interface ISqlTypeFactory {
```

Comment:
Good point. It is only useful for clarifying the concept: `ISqlTypeWith*` extends `ISqlType`. Similarly, `ISqlTypeFactoryWith*` extends `ISqlTypeFactory`.

`ISqlType` and `ISqlTypeFactory` are conceptually "abstract interfaces".

The alternatives were either to make `ISqlType` a union type of all the `ISqlType*` (which I find ugly), or redefine them all as classes, making SqlType and SqlTypeFactory abstract classes.

My main focus with this change was to get the correct return types for the various methods.
If desired, I would happily do a general cleanup later, making (abstract) classes, test, etc.